### ISSUE TITLE
API error change lang

### ISSUE THREAD

**Issue by matthieuowlie** [strapi/strapi#4249]
**Describe the bug**
I wanted my API to send default error messages in french on auth login when the user, associated with identifier and password sent, is not found or blocked.

After having strapi configured (https://strapi.io/documentation/3.0.0-alpha.x/configurations/configurations.html#language) is doesn't worked.

**Steps to reproduce the behavior**
1. Go to  ./config/language.json.
2. Change the language.json defaultLocale with 'fr'
3. Call the API with a POST request on URL "localhost:1337/auth/local" with identifier and password in body.
4. Get the error message in english

**Expected behavior**
The error message should be in the locale language parameter in ./config/language.json

**Screenshots**
![image](https://user-images.githubusercontent.com/56551343/66760064-a52f5a00-eea1-11e9-8ec4-94cb431ba35a.png)

![image](https://user-images.githubusercontent.com/56551343/66760385-46b6ab80-eea2-11e9-8378-5d2710740727.png)

**System**
- Node.js version: v10.16.3
- NPM version: 6.9.0
- Strapi version:  3.0.0-beta.16.5
- Database: sqlite3 in .tmp/data.db
- Operating system: Windows 10

**Comment by lauriejim** [comment]
Hello @matthieuowlie !

The API response messages are only in `en`.
You will not be able to modify them.

I suggest you manage the translation in your front end application.

The configuration you updated is for the backend i18n feature.
That lets you create your translation in `./config/locals` folder.
And use it via the `koa-i18n` middleware.

In the next version of strapi we will remove it.
If you want to use this feature, you will have to manually install it.

**Comment by matthieuowlie** [comment]
Hello @lauriejim ,

thanks for answering my question.

As Strapi does not translate automatically in different languages, is there any way to know what kind of errors Strapi returns ?

In almost all cases, there is a code statusCode 400 with an error 'Bad Request' and only the message sent makes me available to know why it has failed (user has not confirmed his e-mail, user is blocked ...).

But i think doing translation based on a message is not a good practice since the message will maybe change in the future. It would be more efficient to add a code error like "auth/user-not-confirm" "auth/user-blocked"

This way, even if the default message changes, i can trigger a translation or a function depending of the code error returned.

What do you think about at ?

**Comment by lauriejim** [comment]
I agree about the translation based on another translation sentence.
About this thing, we will update to base translation on the error code name (like you said).

To know all auth errors, you can check errors on this files [Auth Controller](https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-users-permissions/controllers/Auth.js)
Check each `formatError` function to see messages.

**Comment by matthieuowlie** [comment]
Hello @lauriejim ,

thanks for sharing all these messages.
From what i saw, maybe you could send the ID of each messages ?

For example, for a user blocked, you send this:
`return ctx.badRequest(
          null,
          formatError({
            id: 'Auth.form.error.blocked',
            message: 'Your account has been blocked by an administrator',
          })
        );
`

Maybe you could send both id and message ? This way i can translate from the id "Auth.form.error.invalid" if necessary or print the default message if not necessary !

**Comment by lauriejim** [comment]
Thank you for this feedback.

**Comment by kuriel-trivu** [comment]
Strapi should return translated messages as @matthieuowlie described. Since the goal of Strapi is to build API with no effort.

![image](https://user-images.githubusercontent.com/46456105/75821497-04fca380-5d64-11ea-8379-c15e35aea477.png)

**Comment by rodrigosebastiao** [comment]
I did a good effort to extract the error message beyond the "400 Bad Request".
In a `.catch(e){ e.networkError.result.errors[0].extensions.data[0].messages[0].message : e.graphQLErrors; }` I am using react, I also did `this.props.error.result.errors[0].extensions.data[0].messages[0].message : e.graphQLErrors;`

A safier way is to traverse the object 
let messageText = findErrorKey(e, "messages");

`function findErrorKey(jsonObj, target) {
                    let level = 0;
                    let response;

                    function traverse(jsonObj) {
                        if(level > 100){ console.log("traverse stack limit", 100); return; }
                        level++;

                        if( jsonObj !== null && typeof jsonObj == "object" ) {
                            Object.entries(jsonObj).forEach(([key, value]) => {
                                // key is either an array index or object key
                                if(key === target){
                                    response = value[0].message;
                                    return response;
                                }
                                traverse(value);
                            });
                        }
                        else {
                            // jsonObj is a number or string
                            // console.log("jsonObj", jsonObj);
                        }
                    }
                    traverse(jsonObj);
                    return response;
                }`

I'd love to know the correct way to forward and get error message in a non-admin frontend and to switch those erros from english to another language.

<img width="733" alt="strapi_error_message" src="https://user-images.githubusercontent.com/25161232/83331372-57b70000-a26c-11ea-89cd-bb39a50b795a.png">

**Comment by rodrigosebastiao** [comment]
I did a good effort to extract the error message beyond the "400 Bad Request".
In a `.catch(e){ e.networkError.result.errors[0].extensions.data[0].messages[0].message : e.graphQLErrors; }` I am using react, I also did `this.props.error.result.errors[0].extensions.data[0].messages[0].message : e.graphQLErrors;`

A safier way is to traverse the object 

let messageText = findErrorKey(e, "messages");

`function findErrorKey(jsonObj, target) {
                    let level = 0;
                    let response;

                    function traverse(jsonObj) {
                        if(level > 100){ console.log("traverse stack limit", 100); return; }
                        level++;

                        if( jsonObj !== null && typeof jsonObj == "object" ) {
                            Object.entries(jsonObj).forEach(([key, value]) => {
                                // key is either an array index or object key
                                if(key === target){
                                    response = value[0].message;
                                    return response;
                                }
                                traverse(value);
                            });
                        }
                        else {
                            // jsonObj is a number or string
                            // console.log("jsonObj", jsonObj);
                        }
                    }
                    traverse(jsonObj);
                    return response;
                }`

I'd love to know the correct way to forward and get error message in a non-admin frontend and to switch those erros from english to another language.

<img width="733" alt="strapi_error_message" src="https://user-images.githubusercontent.com/25161232/83331372-57b70000-a26c-11ea-89cd-bb39a50b795a.png">

**Comment by rodrigosebastiao** [comment]
I did a good effort to extract the error message beyond the "400 Bad Request".
In a 
`.catch(e){ e.networkError.result.errors[0].extensions.data[0].messages[0].message : e.graphQLErrors; }`
I am using react, I also could get from props
`this.props.error.result.errors[0].extensions.data[0].messages[0].message : e.graphQLErrors;`

A safier way is to traverse the object 

`let messageText = findErrorKey(e, "messages");

function findErrorKey(jsonObj, target) {
    let level = 0;
    let response;

    function traverse(jsonObj) {
        if(level > 100){ console.log("traverse stack limit", 100); return; }
        level++;

        if( jsonObj !== null && typeof jsonObj == "object" ) {
            Object.entries(jsonObj).forEach(([key, value]) => {
                // key is either an array index or object key
                if(key === target){
                    response = value[0].message;
                    return response;
                }
                traverse(value);
            });
        }
        else {
            // jsonObj is a number or string
            // console.log("jsonObj", jsonObj);
        }
    }
    traverse(jsonObj);
    return response;
}`

<img width="733" alt="strapi_error_message" src="https://user-images.githubusercontent.com/25161232/83331372-57b70000-a26c-11ea-89cd-bb39a50b795a.png">

I'd love to know the correct way to forward and get error message in a non-admin frontend and to switch those erros from english to another language.

**Comment by ako-v** [comment]
adding ids to the error messages seems to be a need for correct translations, translating errors in the front-end is not a good idea.
I can help with adding ids to the error messages.

**Comment by ldiellyoungl** [comment]
Is there any progress?
